### DISCUSSION TITLE
Failed connection Raspberry Pi with pico through I2C, error occured when send more then one byte data by wire.write()

### DISCUSSION THREAD

**Discussion by bill-rubasse** [earlephilhower/arduino-pico#222] [Q&A] [answered]
Hi, I tried to communicate pico and raspberry pi with I2C, pico assigned as I2C slave, slave address 0x30, raspberry pi as i2c master. An error occurred when I detected i2c address twice:
 
 `sudo i2cdetect -y 1`

First time, raspberry pi is successful detected i2c slave at address 0x30:
![圖片](https://user-images.githubusercontent.com/84432693/123398214-d7c9a980-d5d5-11eb-9e56-ca7f22e1c3c9.png)

Second time, something went wrong:
![圖片](https://user-images.githubusercontent.com/84432693/123398191-d304f580-d5d5-11eb-8ab7-7cae694ce6ec.png)

This is my code on pico:
```
#include <Wire.h>
 byte data[5] = {0x00, 0x01, 0x02, 0x03, 0x04};
void setup() {
  Wire.setSDA(0);
  Wire.setSCL(1);
  Wire.begin(0x30);                // join i2c bus with address #0x30
  Wire.onRequest(requestEvent); // register event
}

void loop() {
  delay(100);
}

void requestEvent() {
  Wire.write(byte, 5); 
  // as expected by master
} 
```
I tried a similar code with normal Arduino nano, this error didn't appear. 
```
#include <Wire.h>

void setup() {
  Wire.begin(0x30);                // join i2c bus with address #0x30
  Wire.onRequest(requestEvent); // register event
}

void loop() {
  delay(100);
}

void requestEvent() {
  Wire.write("hello "); // respond with message of 6 bytes
  // as expected by master
} 
```
![圖片](https://user-images.githubusercontent.com/84432693/123401430-5b38ca00-d5d9-11eb-8ba4-5b9b9eb2817e.png)

Further, I revised my code on pico, sent only one byte data with wire.write() when Wire.onRequest() is called, and it worked. 
```
#include <Wire.h>

void setup() {
  Wire.setSDA(0);
  Wire.setSCL(1);
  Wire.begin(0x30);                // join i2c bus with address #0x30
  Wire.onRequest(requestEvent); // register event
}

void loop() {
  delay(100);
}

void requestEvent() {
  Wire.write(0x01); // respond with message of 1 byte, 0x01
} 
```
I also tried the example code `TalkingToMyself` . In this sample code, Wire.onRequest() sends 6 bytes data through Wire.write(), and it worked.

But further, I tried to connect I2C slave, GPIO2 as SCA, GPIO3 as SCL, with Raspberry Pi, the same error ouucred.
 
Maybe this wire library for pico, will facing error when Wire.onRequest() is called and Wire.write() sends more then one byte. further terminated by i2c master when received NACK?

Is there any solution for the situation, so I can successfully communicate Pi with pico with I2C?

**Comment by Andy2No** [comment]
Do you need to add terminating (pull-up) resistors, for the i2c bus?  Could it just be signal degradation?

**Comment by earlephilhower** [comment, marked as answer]
Good description.  I have an idea what's going on, but need to wire up some stuff to test the theory out.  (Since there are 2 cores and 2 I2C peripherals, I can do your test on the Pico itself).

I believe the issue is that in slave mode you've sent a 5x1-byte writew down but the RPI stops clocking after the 1st byte.  In master mode, because the Pico generates the clock, you can be assured all bytes will eventually go out.  Here, I bet the Pico is stuck on the 2nd byte write which may never exit because the I2C state machine will have reset since the RPI sent I2C end .

**Reply by earlephilhower** [reply]
@bill-rubasse can you give the following patch a try?  It should stop trying to send bytes after a TX_ABRT happens (due to the RPI stopping the read before we're done sending), but I haven't wired up anything to give it a go.

````
diff --git a/libraries/Wire/Wire.cpp b/libraries/Wire/Wire.cpp
index 0cabbc8..df85bb9 100644
--- a/libraries/Wire/Wire.cpp
+++ b/libraries/Wire/Wire.cpp
@@ -315,7 +315,8 @@ size_t TwoWire::write(uint8_t ucData) {
 
 size_t TwoWire::write(const uint8_t *data, size_t quantity) {
     for (size_t i = 0; i < quantity; ++i) {
-        if (!write(data[i])) {
+        // Abort writing if any writes fail or the TX_ABRT flag was set
+        if ((_i2c->hw->intr_stat & (1<<6)) || !write(data[i])) {
             return i;
         }
     }
````

**Reply by bill-rubasse** [reply]
@earlephilhower  Thanks for your reply!

I revised `wire.cpp`,  saved it , and then uploaded `TalkingToMyself` example code to pico again.

![圖片](https://user-images.githubusercontent.com/84432693/123447468-7a048400-d60c-11eb-966b-78bf56c8087b.png)

But the logic analyzer and RPI also showed the same result:

![圖片](https://user-images.githubusercontent.com/84432693/123447234-31e56180-d60c-11eb-84a6-71808f765d62.png)
![圖片](https://user-images.githubusercontent.com/84432693/123447255-390c6f80-d60c-11eb-85ca-7bbc612b00ee.png)

I wasn't sure I did it right, or did it wrong... 
It's my first time trying to revise an arduino library.

**Reply by earlephilhower** [reply]
You did fine, but the problem is evidently something else.  Thanks for trying and reporting back.

I'll need to set up a RPI and the Pico this weekend to give it a go.  I was unable to repro it while using both I2C port on the pico, so there's a slight difference in the RPI i2c detect that I'll need to look at.

**Reply by earlephilhower** [reply]
@bill-rubasse  I was able to exactly reproduce your failure using your description and code.

The issue was related to a TX ABORT flag that was causing the I2C slave state machine to get stuck until cleared (which was never...).  The 1.8.7 release has the 3-line fix in it, and is able to successfully run many of the i2cdetects in a row.

https://github.com/earlephilhower/arduino-pico/commit/aac8146f000e87f88bc742c76af72bcef5118d44

````
pi@retropie:~ $ sudo i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: 30 -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                         
pi@retropie:~ $ sudo i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: 30 -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                         
pi@retropie:~ $ sudo i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:          -- -- -- -- -- -- -- -- -- -- -- -- -- 
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
30: 30 -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- 
70: -- -- -- -- -- -- -- --                         
pi@retropie:~ $ 
````

**Reply by earlephilhower** [reply]
Errr...make that release 1.8.8.  Silly debug message got in there by accident.
